R/utilities-mass-balance.R

```diff
@@ -8,277 +8,238 @@
+    massBalanceData <- getMassBalanceData(
+      groupings = plotSettings$Groupings,
+      compoundNames = compoundNames,
+      simulation = simulation,
+      simulationResults = simulationResults
     )
-  }
-
-  # In the Matlab version, there are multiple groupings performed at that point:
-  # 1) Saliva is excluded from the mass balance calculation
-  # 2) Lumen is a parent compartment that regroups many sub-compartments but Feces
-  # 3) All the rest of the sub-compartments are used normally
 
-  # 1) Exclude Saliva
-  salivaExclusionCondition <- !pathsArray[, "parentCompartmentName"] %in% "Saliva"
-  pathsArray <- pathsArray[salivaExclusionCondition, ]
+    # Get drugMass over time for compoundNames to normalize the mass balance plots
+    applications <- simulation$allApplicationsFor(head(compoundNames, 1))
+    # Get application results need for plot normalization corresponding to data.frame with time and cumulative drugMass
+    applicationResults <- getApplicationResults(applications)
 
-  # 2) Regroup the lumen sub-compartment but Feces
-  lumenGroupingCondition <- pathsArray[, "parentCompartmentName"] %in% "Lumen" & !pathsArray[, "subCompartmentName"] %in% "Feces"
-  pathsArray[lumenGroupingCondition, "subCompartmentName"] <- "Lumen"
-
-  # Factors could not be used for steps 1 and 2 but are needed in the next step
-  pathsArray <- as.data.frame(lapply(pathsArray, function(pathsArrayColumn) {
-    as.factor(pathsArrayColumn)
-  }))
-
-  # Aggregate Data by compound and compartment
-  simulationResultsOutputByGroup <- NULL
-
-  for (compoundLevel in levels(pathsArray[, "compoundName"])) {
-    for (compartmentLevel in levels(pathsArray[, "subCompartmentName"])) {
-      pathFilter <- pathsArray[, "compoundName"] %in% compoundLevel & pathsArray[, "subCompartmentName"] %in% compartmentLevel
-      # as.character() ensures that levels filtered out won't still be added to the aggregation
-      pathNamesFiltered <- as.character(pathsArray[pathFilter, "path"])
+    # Format to appropriate units
+    massBalanceData$Time <- ospsuite::toUnit(
+      "Time",
+      massBalanceData$Time,
+      structureSet$simulationSet$timeUnit
+    )
+    timeTicks <- getTimeTicksFromUnit(
+      structureSet$simulationSet$timeUnit,
+      timeValues = massBalanceData$Time
+    )
 
-      # cbind(..., 0) prevents rowSums to crash if only one column is filtered
-      aggregatedData <- rowSums(cbind.data.frame(
-        simulationResultsOutput$data[, pathNamesFiltered, drop = FALSE],
-        data.frame(dummyVariable = 0)
-      ))
+    metaData <- list(
+      "Time" = list(
+        dimension = "Time",
+        unit = structureSet$simulationSet$timeUnit
+      ),
+      "Amount" = list(
+        dimension = applications[[1]]$drugMass$dimension,
+        unit = applications[[1]]$drugMass$unit
+      ),
+      "NormalizedAmount" = list(
+        dimension = applications[[1]]$drugMass$dimension,
+        unit = "fraction of drugmass"
```

Comment:
dimension should be "Fraction" here and not `drugMass$dimension`, right?

Reply:
Here, I am simply leveraging the smart mapping from tlf that will display `dimension [unit]` in the plot.
This way we get `Amount [fraction of drugmass]` displayed in the plot axis.
Let me know if you wish to display the normalized plot differently

Reply:
ok, let's keep for now like this :)